internal/pkg/manifest/storage.go

```diff
@@ -28,13 +38,87 @@ type SidecarMountPoint struct {
 
 // EFSVolumeConfiguration holds options which tell ECS how to reach out to the EFS filesystem.
 type EFSVolumeConfiguration struct {
-	FileSystemID  *string              `yaml:"id"`       // Required.
-	RootDirectory *string              `yaml:"root_dir"` // Default "/"
-	AuthConfig    *AuthorizationConfig `yaml:"auth"`
+	FileSystemID  *string              `yaml:"id"`       // Required. Can be specified as "copilot" or "managed" magic keys.
+	RootDirectory *string              `yaml:"root_dir"` // Default "/". For BYO EFS.
+	AuthConfig    *AuthorizationConfig `yaml:"auth"`     // Auth config for BYO EFS.
+	UID           *uint32              `yaml:"uid"`      // UID for managed EFS.
+	GID           *uint32              `yaml:"gid"`      // GID for managed EFS.
+}
+
+// IsEmpty returns empty if the struct has all zero members.
+func (e *EFSVolumeConfiguration) IsEmpty() bool {
+	if e.FileSystemID == nil && e.RootDirectory == nil && e.AuthConfig == nil && e.UID == nil && e.GID == nil {
+		return true
+	}
+	return false
+}
+
+// EFSConfigOrID contains custom unmarshaling logic for the `efs` field in the manifest.
+type EFSConfigOrID struct {
+	Config EFSVolumeConfiguration
+	ID     string
+}
+
+// UnmarshalYAML implements the yaml(v2) interface. It allows EFS to be specified as a
+// string or a struct alternately.
+func (e *EFSConfigOrID) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	if err := unmarshal(&e.Config); err != nil {
+		switch err.(type) {
+		case *yaml.TypeError:
+			break
+		default:
+			return err
+		}
+	}
+
+	if !e.Config.IsEmpty() {
+		// Unmarshaled successfully to e.Config, unset e.ID, and return.
+		e.ID = ""
+		return nil
+	}
+
+	if err := unmarshal(&e.ID); err != nil {
+		return errUnmarshalBuildOpts
+	}
+	return nil
+}
+
+// UseManagedFS returns true if the user has specified "copilot" or "managed" as a FSID; false otherwise.
+func (e *EFSConfigOrID) UseManagedFS() bool {
+	if contains(managedFSIDKeys, e.ID) {
+		return true
+	}
+	if contains(managedFSIDKeys, aws.StringValue(e.Config.FileSystemID)) {
+		return true
+	}
+	return false
+}
+
+// FSID returns the correct value of the EFS filesystem ID. If the ID is set improperly (via a bad merge
+// of environment overrides), it throws an error.
+func (e *EFSConfigOrID) FSID() (*string, error) {
+	fromID := e.ID
+	if e.Config.IsEmpty() {
+		return aws.String(fromID), nil
+	}
+	fromConfig := aws.StringValue(e.Config.FileSystemID)
+	if fromID != "" && fromID != fromConfig {
+		return nil, fmt.Errorf("read EFS ID: multiple values specified (%s, %s)", fromID, fromConfig)
```

Comment:
Can we just ignore `e.ID` if `e.Config` is not empty? I remember we did the very similar thing with the other custom unmarshal.